test/xbgpu/test_engine.py

```diff
@@ -1086,8 +1132,14 @@ def heap_factory(batch_index: int, present: np.ndarray) -> list[spead2.send.Heap
                 present[40:50, missing_antenna] = False  # Covers some complete accumulations
                 present[60, missing_antenna] = False  # Just one heap in an accumulation
 
-            for beam_request in flattened_requests:
+            for beam_request in list(chain.from_iterable(katcp_requests[:N_BEAM_REQUESTS])):
```

Comment:
```suggestion
            for beam_request in chain.from_iterable(katcp_requests[:N_BEAM_REQUESTS]):
```
You can iterate directly over `from_iterable` without collecting it into a list first.

Reply:
Again, the indexing looks fishy.